x/reward/keeper/reward_test.go

```diff
@@ -0,0 +1,233 @@
+		{
+			name: "valid case",
+			args: args{
+				blockRatio: sdk.NewDecWithPrec(6, 1),
+				ratio:      sdk.NewDecWithPrec(34, 2),
+				coins:      coins,
+			},
+			want: sdk.NewCoins(
+				sdk.NewCoin(coinA.Denom, sdk.NewInt(7960000)),
+				sdk.NewCoin(coinB.Denom, sdk.NewInt(2388)),
+				sdk.NewCoin(coinC.Denom, sdk.NewInt(8)),
+			),
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := keeper.CalculateReward(tt.args.blockRatio, tt.args.ratio, tt.args.coins)
+			if tt.wantErr {
+				require.Error(t, err)
+				return
+			}
+			require.NoError(t, err)
+			require.ElementsMatch(t, tt.want, got)
+		})
+	}
+}
+
+func TestKeeper_DistributeRewards(t *testing.T) {
+	var (
+		k, _, pk, bk, _, _, _, ctx = setupMsgServer(t)
+		launchIDs                  = []uint64{1, 2, 3, 4, 5}
+		validatorFoo               = sample.Address()
+		validatorBar               = sample.Address()
+		validatorConsAddrFoo       = sample.Address()
+		validatorConsAddrBar       = sample.Address()
+		validatorConsAddrBaz       = sample.Address()
+		notFoundValidatorAddr      = sample.Address()
+		provider                   = sample.Address()
+		coins                      = sample.Coins()
+		signatureCounts            = spntypes.SignatureCounts{
+			BlockCount: 2,
+			Counts: []spntypes.SignatureCount{
+				{ConsAddress: validatorConsAddrFoo, RelativeSignatures: sdk.NewDec(1)},
+				{ConsAddress: validatorConsAddrBar, RelativeSignatures: sdk.NewDecWithPrec(5, 1)},
+				{ConsAddress: validatorConsAddrBaz, RelativeSignatures: sdk.NewDecWithPrec(55, 1)},
+			},
+		}
+		signatureCountsValNotFound = signatureCounts
+	)
+	signatureCountsValNotFound.Counts = append(signatureCountsValNotFound.Counts, spntypes.SignatureCount{
+		ConsAddress: notFoundValidatorAddr, RelativeSignatures: sdk.NewDec(1),
+	})
+	moduleSupply := coins
+	for _, launchID := range launchIDs {
+		k.SetRewardPool(ctx, types.RewardPool{
+			LaunchID:            launchID,
+			Provider:            provider,
+			Coins:               coins,
+			LastRewardHeight:    1,
+			CurrentRewardHeight: 4,
+		})
+		moduleSupply = moduleSupply.Add(moduleSupply...)
+	}
+	pk.SetValidator(ctx, profiletypes.Validator{
+		Address:          validatorFoo,
+		ConsensusAddress: validatorConsAddrFoo,
+	})
+	pk.SetValidatorByConsAddress(ctx, profiletypes.ValidatorByConsAddress{
+		ValidatorAddress: validatorFoo,
+		ConsensusAddress: validatorConsAddrFoo,
+	})
+	pk.SetValidator(ctx, profiletypes.Validator{
+		Address:          validatorBar,
+		ConsensusAddress: validatorConsAddrBar,
+	})
+	pk.SetValidatorByConsAddress(ctx, profiletypes.ValidatorByConsAddress{
+		ValidatorAddress: validatorBar,
+		ConsensusAddress: validatorConsAddrBar,
+	})
+	pk.SetValidatorByConsAddress(ctx, profiletypes.ValidatorByConsAddress{
+		ValidatorAddress: sample.Address(),
+		ConsensusAddress: notFoundValidatorAddr,
+	})
+	err := bk.MintCoins(ctx, types.ModuleName, moduleSupply)
+	require.NoError(t, err)
+
+	type args struct {
```

Comment:
We should assert the coin distribution in the tests
We could for example provide in the input an array `[]Distributed{address: string, coins: string}` and check the balance for accounts
We can eventually initialize basic cases for the PR and handle extensive cases in future PRs since this part is specifically critical security speaking
We can also asser the balance of the provider for refund

Reply:
done one https://github.com/tendermint/spn/pull/488/commits/f5d2c7392e282469e0c61c8d857767d181f30dfb
but we can improve more these test cases later